crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +299,385 @@ where
+        // as this message and note its sequence number
+        for key in keys_changed {
+            let Some(path) = is_ibc_commitment_key(key) else {
+                continue;
+            };
+            if path.port_id == message.port_id_on_a
+                && path.channel_id == message.chan_id_on_a
+            {
+                let Some(storage_commitment): Option<PacketCommitment> =
+                    self.ctx.read_bytes_post(key)?.map(Into::into)
+                else {
+                    // Ignore this event if it does not exist
+                    continue;
+                };
+                if packet_commitment == storage_commitment {
+                    return Ok(Some(path.sequence));
+                }
+            }
+        }
+        Ok(None)
+    }
+
+    // Try to determine which address would cause query_ibc_denom to yield the
+    // supplied denom
+    fn reverse_query_ibc_denom(
+        denom: &PrefixedDenom,
+        ibc_denoms: &BTreeMap<String, Address>,
+    ) -> Option<Address> {
+        ibc_denoms
+            .get(&denom.to_string())
+            .cloned()
+            // If the reverse lookup failed, then guess the Address
+            // that might have yielded the IBC denom. However,
+            // guessing an IBC token address cannot possibly be
+            // correct due to the structure of query_ibc_denom
+            .or_else(|| {
+                Address::decode(denom.to_string()).ok().filter(|x| {
+                    !matches!(
+                        x,
+                        Address::Internal(InternalAddress::IbcToken(_))
+                    )
+                })
+            })
+    }
+
+    // Apply the given send packet to the changed balances structure
+    fn apply_send_packet(
+        &self,
+        mut acc: ChangedBalances,
+        msg: &MsgTransfer,
+        keys_changed: &BTreeSet<Key>,
+    ) -> Result<ChangedBalances> {
+        // If a key change with the same port, channel, and commitment as this
+        // message cannot be found, then ignore this message
+        if self
+            .search_ibc_transfer(&msg.message, keys_changed)?
+            .is_none()
+        {
+            return Ok(acc);
+        };
+
+        // Since IBC denominations are derived from Addresses
+        // when sending, we have to do a reverse look-up of the
+        // relevant token Address
+        let Some(token) = Self::reverse_query_ibc_denom(
+            &msg.message.packet_data.token.denom,
+            &acc.ibc_denoms,
+        ) else {
+            return Ok(acc);
+        };
+        let delta = ValueSum::from_pair(
+            token.clone(),
+            Amount::from_uint(Uint(*msg.message.packet_data.token.amount), 0)
+                .unwrap(),
+        );
+        // If there is a transfer to the IBC account, then deduplicate the
+        // balance increase since we already accounted for it above
```

Comment:
What do you mean "above" ?

Reply:
My mistake, the computation is now below. See here: https://github.com/anoma/namada/blob/08e56643fd65d42de3cf854122ab1fe74698f7ac/crates/namada/src/ledger/native_vp/masp.rs#L451 . Will fix this.